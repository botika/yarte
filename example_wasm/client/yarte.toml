# TODO: unique config file or composable
[main]
dir = "../templates"
debug = "code"

[debug]
short = false
