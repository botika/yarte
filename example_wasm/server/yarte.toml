[main]
dir = "../templates"

[debug]
short = false
