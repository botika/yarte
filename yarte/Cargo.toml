[package]
name = "yarte"
version = "0.5.1"
authors = ["Juan Aguilar Santillana <mhpoin@gmail.com>"]
description = "Type-safe, compiled Handlebars"
categories = ["template-engine"]
documentation = "https://docs.rs/yarte"
edition = "2018"
include = ["src/**/*", "Cargo.toml"]
keywords = ["markup", "template", "handlebars", "html"]
license = "MIT/Apache-2.0"
readme = "../README.md"
repository = "https://github.com/botika/yarte"
workspace = ".."

[badges]
travis-ci = { repository = "botika/yarte", branch = "master" }
maintenance = { status = "actively-developed" }

[features]
default = []
with-actix-web = ["actix-web", "futures", "yarte_derive/actix-web"]
full = ["with-actix-web"]

[dependencies]
yarte_config = { version = "0.5", path = "../yarte_config" }
yarte_derive = { version = "0.5", path = "../yarte_derive" }
yarte_helpers = { version = "0.5", path = "../yarte_helpers" }

actix-web = { version = "2.0", optional = true }
futures = { version = "0.3", optional = true }

yarte_wasm_app = { version = "0.0", path = "../yarte_wasm_app" }

[build-dependencies]
yarte_config = { version = "0.5", path = "../yarte_config" }
