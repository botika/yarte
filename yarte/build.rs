use yarte_helpers::recompile::when_changed;

fn main() {
    when_changed();
}
