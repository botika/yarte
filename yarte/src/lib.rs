//! Yarte stands for **Y**et **A**nother **R**ust **T**emplate **E**ngine,
//! is the fastest template engine. Uses a Handlebars-like syntax,
//! well known and intuitive for most developers. Yarte is an optimized, and easy-to-use
//! rust crate, can create logic around their templates using using conditionals, loops,
//! rust code and templates composition.
//!
//! Also Yarte incorporates feature `with-actix-web`, an implementation of `actix-web`'s
//! trait Responder for those using this framework.
//!
//! [Yarte book](https://yarte.netlify.com)
//!

#[cfg(not(target_arch = "wasm32"))]
use std::fmt::{self, Write};

pub use yarte_derive::Template;
#[cfg(not(target_arch = "wasm32"))]
pub use yarte_helpers::{helpers::Render, Error, Result};
#[cfg(target_arch = "wasm32")]
pub use yarte_wasm_app::App as Template;

pub mod recompile;

/// Template trait, will implement by derive like `Display` or `actix_web::Responder` (with feature)
///
/// ```rust
/// use yarte::Template;
///
/// #[derive(Template)]
/// #[template(src="Hello, {{ name }}!")]
/// struct HelloTemplate<'a> {
///     name: &'a str,
/// }
///
/// println!("{}", HelloTemplate { name: "world" })
/// ```
///
#[cfg(not(target_arch = "wasm32"))]
pub trait Template: fmt::Display {
    /// which will write this template
    fn call(&self) -> Result<String> {
        let mut buf = String::with_capacity(Self::size_hint());
        write!(buf, "{}", self).map(|_| buf)
    }

    /// https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
    #[cfg(feature = "with-actix-web")]
    fn mime() -> &'static str
    where
        Self: Sized;

    /// Approximation of output size used in method `call`.
    /// Yarte implements an heuristic algorithm of allocation.
    fn size_hint() -> usize;
}

#[cfg(all(feature = "with-actix-web", not(target_arch = "wasm32")))]
pub mod aw {
    pub use actix_web::{
        error::ErrorInternalServerError, Error, HttpRequest, HttpResponse, Responder,
    };
    pub use futures::future::{err, ok, Ready};
}

#[cfg(all(feature = "wasm", not(target_arch = "wasm32")))]
pub mod serde_json {
    pub use serde_json::to_string;
}
