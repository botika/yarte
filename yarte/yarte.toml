# root dir of templates
[main]
dir = "templates"

# Alias for partials. In call, change the start of partial path with one of this, if exist.
[partials]
alias = "example/deep/more/deep"
doc = "example/deep/more/doc"
card = "example/deep/more/card"
