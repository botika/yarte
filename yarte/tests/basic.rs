#![allow(clippy::blacklisted_name)]

use std::collections::HashMap;

use yarte::Template;

#[derive(Template)]
#[template(path = "hello.hbs")]

struct HelloTemplate<'a> {
    name: &'a str,
}

#[test]
fn test_hello() {
    let hello = HelloTemplate { name: "world" };
    assert_eq!("Hello, world!", hello.call().unwrap());
}

#[derive(Template)]
#[template(src = "{{}", ext = "txt")]

struct BracketsTemplate;

#[test]
fn test_brackets() {
    let hello = BracketsTemplate;
    assert_eq!("{{}", hello.call().unwrap());
}

#[derive(Template)]
#[template(src = "{{{}}", ext = "txt")]

struct Brackets2Template;

#[test]
fn test_brackets2() {
    let hello = Brackets2Template;
    assert_eq!("{{{}}", hello.call().unwrap());
}
#[derive(Template)]
#[template(path = "simple.hbs")]
struct VariablesTemplate<'a> {
    strvar: &'a str,
    num: i64,
    i18n: String,
}

#[test]
fn test_variables() {
    let s = VariablesTemplate {
        strvar: "foo",
        num: 42,
        i18n: "Iñtërnâtiônàlizætiøn".to_string(),
    };
    assert_eq!(
        s.call().unwrap(),
        "hello world, foo\n\
         with number: 42\n\
         Iñtërnâtiônàlizætiøn is important\n\
         in vars too: Iñtërnâtiônàlizætiøn"
    );
    assert_eq!(VariablesTemplate::mime(), "text/html; charset=utf-8");
}

#[derive(Template)]
#[template(path = "hello.hbs")]
struct EscapeTemplate<'a> {
    name: &'a str,
}

#[test]
fn test_escape() {
    let s = EscapeTemplate { name: "<>&\"'/" };

    assert_eq!(s.call().unwrap(), "Hello, &lt;&gt;&amp;&quot;&#x27;&#x2f;!");
}

#[derive(Template)]
#[template(path = "simple-no-escape.txt")]
struct VariablesTemplateNoEscape<'a> {
    strvar: &'a str,
    num: i64,
    i18n: String,
}

#[test]
fn test_variables_no_escape() {
    let s = VariablesTemplateNoEscape {
        strvar: "foo",
        num: 42,
        i18n: "Iñtërnâtiônàlizætiøn".to_string(),
    };
    assert_eq!(
        s.call().unwrap(),
        "hello world, foo\n\
         with number: 42\n\
         Iñtërnâtiônàlizætiøn is important\n\
         in vars too: Iñtërnâtiônàlizætiøn"
    );
}

#[derive(Template)]
#[template(path = "if.hbs")]
struct IfTemplate {
    cond: bool,
}

#[test]
fn test_if() {
    let s = IfTemplate { cond: true };
    assert_eq!(s.call().unwrap(), "true");
}

#[derive(Template)]
#[template(path = "else.hbs")]
struct ElseTemplate {
    cond: bool,
}

#[test]
fn test_else_false() {
    let s = ElseTemplate { cond: false };
    assert_eq!(s.call().unwrap(), "     \n    false\n");
}

#[test]
fn test_else_true() {
    let s = ElseTemplate { cond: true };
    assert_eq!(s.call().unwrap(), "     \n true");
}

#[derive(Template)]
#[template(path = "else-if.hbs")]
struct ElseIfTemplate {
    cond: bool,
    check: bool,
}

#[test]
fn test_else_if() {
    let s = ElseIfTemplate {
        cond: false,
        check: true,
    };
    assert_eq!(s.call().unwrap(), " checked ");
}

#[derive(Template)]
#[template(path = "comment.hbs")]
struct CommentTemplate {}

#[test]
fn test_comment() {
    let t = CommentTemplate {};
    assert_eq!(t.call().unwrap(), "");
}

#[derive(Template)]
#[template(path = "negation.hbs")]
struct NegationTemplate {
    foo: bool,
}

#[test]
fn test_negation() {
    let t = NegationTemplate { foo: false };
    assert_eq!(t.call().unwrap(), "Hello");
}

#[derive(Template)]
#[template(path = "minus.hbs")]
struct MinusTemplate {
    foo: i8,
}

#[test]
fn test_minus() {
    let t = MinusTemplate { foo: 1 };
    assert_eq!(t.call().unwrap(), "Hello");
}

#[derive(Template)]
#[template(path = "index.hbs")]
struct IndexTemplate {
    foo: HashMap<String, String>,
}

#[test]
fn test_index() {
    let mut foo = HashMap::new();
    foo.insert("bar".into(), "baz".into());
    let t = IndexTemplate { foo };
    assert_eq!(t.call().unwrap(), "baz");
}

#[derive(Template)]
#[template(path = "tuple-attr.hbs")]
struct TupleAttrTemplate<'a>(&'a str, &'a str);

#[test]
fn test_tuple_attr() {
    let t = TupleAttrTemplate("foo", "bar");
    assert_eq!(t.call().unwrap(), "foobar");
}

struct Holder {
    a: usize,
}

struct NestedHolder {
    holder: Holder,
}

#[derive(Template)]
#[template(path = "nested-attr.hbs")]
struct NestedAttrTemplate {
    inner: NestedHolder,
}

#[test]
fn test_nested_attr() {
    let t = NestedAttrTemplate {
        inner: NestedHolder {
            holder: Holder { a: 5 },
        },
    };
    assert_eq!(t.call().unwrap(), "5");
}

#[derive(Template)]
#[template(path = "literals.hbs")]
struct LiteralsTemplate {}

#[test]
fn test_literals() {
    let s = LiteralsTemplate {};
    assert_eq!(s.call().unwrap(), "a");
}

#[derive(Template)]
#[template(path = "empty.hbs")]
struct Empty;

#[test]
fn test_empty() {
    assert_eq!(Empty.call().unwrap(), "foo");
}

struct Foo {
    a: usize,
}

#[derive(Template)]
#[template(path = "attr.hbs")]
struct AttrTemplate {
    inner: Foo,
}

#[test]
fn test_attr() {
    let t = AttrTemplate {
        inner: Foo { a: 1 },
    };
    assert_eq!(t.call().unwrap(), "1");
}

#[derive(Template)]
#[template(path = "option.hbs")]
struct OptionTemplate {
    var: Option<usize>,
}

#[test]
fn test_option() {
    let t = OptionTemplate { var: Some(1) };
    assert_eq!(t.call().unwrap(), "some: 1");
}

#[derive(Template)]
#[template(src = "{{{ string }}}")]
struct Unwrapped {
    string: String,
}

#[test]
fn test_unwrapped() {
    let t = Unwrapped {
        string: String::from("&"),
    };
    assert_eq!("&", t.call().unwrap());
}
