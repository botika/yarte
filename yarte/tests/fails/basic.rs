use yarte::Template;

#[derive(Template)]
#[template(src = "{{ @foo }}")]
struct Test{
    foo: usize
}

fn main() {}