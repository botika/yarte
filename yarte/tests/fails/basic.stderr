error: proc-macro derive panicked
 --> $DIR/basic.rs:3:10
  |
3 | #[derive(Template)]
  |          ^^^^^^^^
  |
  = help: message: error
           --> $DIR/Test.hbs:1:4
            |
          1 | {{ @foo }}
            |    ^ expression expected expression
            |
